### PR TITLE
ipq40xx: add support for GL.iNet GL-S1300

### PR DESCRIPTION
**Pull request by handongming**
Specifications:
SOC:        Qualcomm IPQ4029 (DAKOTA) ARM Quad-Core
RAM:        512 MiB
FLASH1:     16 MiB NOR - SPI0
FLASH2:     8 GiB eMMC
ETH:        Qualcomm QCA8075
WLAN1:      Qualcomm Atheros QCA4029 2.4GHz 802.11b/g/n 2x2
WLAN2:      Qualcomm Atheros QCA4029 5GHz 802.11n/ac W2 2x2
INPUT:      Reset, WPS
LED:        Power, Mesh, WLAN
UART1:      On board pin header near to LED (3.3V, TX, RX, GND), 3.3V without pin - 115200 8N1
UART2:      On board with BLE module
SPI1:       On board socket for Zigbee module

Install via tftp
- NB: need to flash transition image firstly

Firstly install transition image:
(IPQ40xx) # tftpboot 0x84000000 s1300-factory-to-openwrt.img
(IPQ40xx) # sf probe && imgaddr=0x84000000 && source :script

Secondly install openwrt sysupgrade bin:
(IPQ40xx) # run lf

Revert to factory image:
(IPQ40xx) # tftpboot 0x84000000 s1300-openwrt-to-factory.img
(IPQ40xx) # sf probe && imgaddr=0x84000000 && source :script

The kernel and rootfs of factory firmware are on eMMC, and openwrt
firmware is on NOR flash. The transition image includes U-boot
and partition table, which decides where to load kernel and rootfs.
After you firstly install openwrt image, you can switch between
factory and openwrt firmware by flashing transition image.

Tested okay with: WiFi 2.4G and 5G, USB storage, UART0 for serial, UART1 for BLE, SPI for zigbee, ethernet WAN and LAN.

### CODE REVIEW

File: package/firmware/ipq-wifi/Makefile [hunk 1]
```diff
@@ -33,7 +33,8 @@ ALLWIFIBOARDS:= \
 	engenius_emd1 \
 	ezviz_cs-w3-wd1200g-eup \
 	linksys_ea8300 \
-	qxwlan_e2600ac
+	qxwlan_e2600ac \
+	glinet_gl-s1300
```
Comment (CHKDSK88) [inline package/firmware/ipq-wifi/Makefile:37]: You should keep alphabetical order.

File: package/firmware/ipq-wifi/Makefile [hunk 2]
```diff
@@ -103,5 +104,6 @@ $(eval $(call generate-ipq-wifi-package,engenius_emd1,EnGenius EMD1))
 $(eval $(call generate-ipq-wifi-package,ezviz_cs-w3-wd1200g-eup,EZVIZ CS-W3-WD1200G EUP))
 $(eval $(call generate-ipq-wifi-package,linksys_ea8300,Linksys EA8300))
 $(eval $(call generate-ipq-wifi-package,qxwlan_e2600ac,Qxwlan E2600AC))
+$(eval $(call generate-ipq-wifi-package,glinet_gl-s1300,GL.iNet GL-S1300))
```
Comment (CHKDSK88) [inline package/firmware/ipq-wifi/Makefile:107]: You should keep alphabetical order.

File: target/linux/ipq40xx/base-files/etc/board.d/01_leds [hunk 3]
```diff
@@ -24,6 +24,7 @@ avm,fritzbox-4040)
 	ucidef_set_led_switch "lan" "LAN" "fritz4040:green:lan" "switch0" "0x1e"
 	;;
 avm,fritzbox-7530 |\
+glinet,gl-s1300 |\
```
Comment (CHKDSK88) [inline target/linux/ipq40xx/base-files/etc/board.d/01_leds:27]: You should keep alphabetical order.

File: target/linux/ipq40xx/base-files/etc/board.d/02_network [hunk 4]
```diff
@@ -68,6 +68,7 @@ ipq40xx_setup_interfaces()
 		ucidef_add_switch "switch0" \
 			"0u@eth0" "2:lan:3" "3:lan:2" "4:lan:1" "0u@eth1" "5:wan"
 		;;
+	glinet,gl-s1300 |\
```
Comment (CHKDSK88) [inline target/linux/ipq40xx/base-files/etc/board.d/02_network:71]: You should keep alphabetical order.

File: target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata [hunk 5]
```diff
@@ -49,6 +49,7 @@ case "$FIRMWARE" in
 	alfa-network,ap120c-ac |\
 	cilab,meshpoint-one |\
 	ezviz,cs-w3-wd1200g-eup |\
+	glinet,gl-s1300 |\
```
Comment (CHKDSK88) [inline target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:52]: You should keep alphabetical order.

File: target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata [hunk 6]
```diff
@@ -130,6 +131,7 @@ case "$FIRMWARE" in
...
```
Comment (CHKDSK88) [inline target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:134]: You should keep alphabetical order.

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4029-gl-s1300.dts [hunk 7]
```diff
@@ -0,0 +1,368 @@
+/* Copyright (c) 2015, The Linux Foundation. All rights reserved.
+ *
```
Comment (CHKDSK88) [inline target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4029-gl-s1300.dts:2]: Change to // SPDX-License-Identifier?

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4029-gl-s1300.dts [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+			label = "reset";
+			gpios = <&tlmm 18 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		power: power {
```
Comment (CHKDSK88) [inline target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4029-gl-s1300.dts:123]: led_power: power alias?

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4029-gl-s1300.dts [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+	status = "okay";
+};
+
+&blsp1_spi1 {
+	pinctrl-0 = <&spi_0_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+	cs-gpios = <&tlmm 12 GPIO_ACTIVE_HIGH>;
+
+	mx25l25635f@0 {
```
Comment (CHKDSK88) [inline target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4029-gl-s1300.dts:166]: flash@0

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4029-gl-s1300.dts [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+				reg = <0x180000 0xe80000>;
+			};
+		};
+	};
+};
+
+&blsp1_spi2 {
+	pinctrl-0 = <&spi_1_pins>;
+	pinctrl-names = "default";
+	status = "ok";
```
Comment (CHKDSK88) [inline target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4029-gl-s1300.dts:236]: ok is depreciated. Use "okay" instead.

**Comment by handongming** [comment]
@CHKDSK88 Thanks for review, updated.

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4029-gl-s1300.dts [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+		wlan {
+			label = "gl-s1300:green:wlan";
+			gpios = <&tlmm 60 GPIO_ACTIVE_HIGH>;
+		};
+	};
+};
+
+&vqmmc {
+	status = "okay";
+};
```
Comment (adrianschmutzler) [inline target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4029-gl-s1300.dts:129]: Missing empty line after node.

File: target/linux/ipq40xx/files-5.4/arch/arm/boot/dts/qcom-ipq4029-gl-s1300.dts [hunk 12]
```diff
@@ -0,0 +1,354 @@
...
+		wlan {
+			label = "gl-s1300:green:wlan";
+			gpios = <&tlmm 60 GPIO_ACTIVE_HIGH>;
+		};
+	};
+};
+
+&vqmmc {
+	status = "okay";
+};
```
Comment (adrianschmutzler) [inline target/linux/ipq40xx/files-5.4/arch/arm/boot/dts/qcom-ipq4029-gl-s1300.dts:129]: Missing empty line after node.

File: target/linux/ipq40xx/image/Makefile [hunk 13]
```diff
@@ -387,6 +387,20 @@ define Device/glinet_gl-b1300
 endef
 TARGET_DEVICES += glinet_gl-b1300
 
+define Device/glinet_gl-s1300
+	$(call Device/FitImage)
+	DEVICE_VENDOR := GL.iNet
+	DEVICE_MODEL := GL-S1300
+	BOARD_NAME := gl-s1300
```
Comment (adrianschmutzler) [inline target/linux/ipq40xx/image/Makefile:394]: Is this really needed?

Comment (handongming) [reply target/linux/ipq40xx/image/Makefile:394]: I found if delete these lines, the menuconfig won't show the target profile, so it's needed.

Comment (adrianschmutzler) [reply target/linux/ipq40xx/image/Makefile:394]: I was referring to the BOARD_NAME only. The DEVICE_* stuff is indeed needed for menuconfig.

Comment (handongming) [reply target/linux/ipq40xx/image/Makefile:394]: Thanks. I check BOARD_NAME is only included in sysupgrade.meta for variable "board", and that's not used for sysupgrade, so remove it is okay as you mention.

**Comment by pniaps** [comment]
What is needed for this to merge?. Can I help or is @handongming who has to fix this?

Thanks

File: target/linux/ipq40xx/base-files/etc/board.d/01_leds [hunk 14]
```diff
@@ -24,7 +24,8 @@ avm,fritzbox-4040)
 	ucidef_set_led_switch "lan" "LAN" "fritz4040:green:lan" "switch0" "0x1e"
 	;;
 avm,fritzbox-7530 |\
-glinet,gl-b1300)
+glinet,gl-b1300 |\
+glinet,gl-s1300)
 	ucidef_set_led_wlan "wlan" "WLAN" "${boardname}:green:wlan" "phy0tpt"
```
Comment (adrianschmutzler) [inline target/linux/ipq40xx/base-files/etc/board.d/01_leds:29]: Can you try with trigger in DTS instead?

File: target/linux/ipq40xx/image/Makefile (hunk 13 above)
Comment (adrianschmutzler) [reply target/linux/ipq40xx/image/Makefile:394]: So, remove it?

File: target/linux/ipq40xx/image/Makefile [hunk 15]
```diff
@@ -387,6 +387,20 @@ define Device/glinet_gl-b1300
...
+	SOC := qcom-ipq4029
+	KERNEL_SIZE := 4096k
+	IMAGE_SIZE := 26624k
+	IMAGES := sysupgrade.bin
+	IMAGE/sysupgrade.bin := append-kernel |append-rootfs | pad-rootfs | append-metadata
```
Comment (adrianschmutzler) [inline target/linux/ipq40xx/image/Makefile:399]: Missing space after first pipe.

**Comment by adrianschmutzler** [comment]
I've just merged the dts folder for 4.19 and 5.4 to a single one in commit fa6d53261a68997a7928f38578c5824a2d375d6f, so you may adjust your PR accordingly (i.e. keep only one proper DTS).

**Comment by handongming** [comment]
Thanks, updated.

**Comment by pniaps** [comment]
I've tested this and works. Thanks.

**Comment by ynezz** [comment]
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git

**Comment by pniaps** [comment]
When will this device in a stable relase? 19.07.4 o maybe 20.x?

Perhaps I can contribute, but I don't know how to do it.

**Comment by adschm** [comment]
Since this device has been merged to master already, it will be in the _next_ stable release, i.e. 20.xx.

**Pull request closed without merging**